lib/services/role.go

```diff
@@ -1428,141 +1416,119 @@ func (set RoleSet) CheckAccessToKubernetes(namespace string, kube *KubernetesClu
 	return trace.AccessDenied("access to kubernetes cluster denied")
 }
 
-// CheckAccessToDatabaseServer checks if this role set has access to the
-// specified database server.
-//
-// Used to filter available databases a user sees with "tsh db ls" command.
-func (set RoleSet) CheckAccessToDatabaseServer(server types.DatabaseServer) error {
-	var errs []error
-	// Check deny rules.
-	for _, role := range set {
-		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Deny), server.GetNamespace())
-		matchLabels, labelsMessage, err := MatchLabels(role.GetDatabaseLabels(Deny), server.GetAllLabels())
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if matchNamespace && matchLabels {
-			log.WithField(trace.Component, teleport.ComponentRBAC).Debugf(
-				"Access to database %q denied, deny rule in %q matched; match(namespace=%v, label=%v).",
-				server.GetName(), role.GetName(), namespaceMessage, labelsMessage)
-			return trace.AccessDenied("access to database denied")
-		}
-	}
-	// Check allow rules.
-	for _, role := range set {
-		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Allow), server.GetNamespace())
-		matchLabels, labelsMessage, err := MatchLabels(role.GetDatabaseLabels(Allow), server.GetAllLabels())
+// Matcher defines an interface for a generic role matcher.
+type Matcher interface {
```

Comment:
`RoleMatcher` since this is inside a generic services package and might be confusing.